+ added keyboard support

1lyyfe opened the pull request:
+ added keyboard support for viewControllers which conform to shortFo…

+ update scrollview position when showing keyboard if needed

###  Summary

Describe the goal of this PR. Mention any related Issue numbers.

### Requirements (place an `x` in each `[ ]`)

* [ ] I've read and understood the [Contributing Guidelines](https://github.com/slackhq/PanModal/blob/master/CONTRIBUTING.md) and have done my best effort to follow them.
* [ ] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

* [ ] I've written tests to cover the new code and functionality included in this PR.

1lyyfe committed:
+ added keyboard support for viewControllers which conform to shortForm and longForm states
+ update scrollview position when showing keyboard if needed

1lyyfe committed:
+ removed hardcoded value

markdarling commented:
Should you not be de-observing the keyboard somewhere?

markj commented on PanModal/Controller/PanModalPresentationController.swift:
```diff
@@ -234,12 +246,57 @@ open class PanModalPresentationController: UIPresentationController {
+        
+        //if the presented view is already at max height and is scrollable then scroll to above the keyboard when keyboard is shown if needed
+        
+        if let keyboardSize = (notification.userInfo?  [UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue {
```
That's a bit crazy.. why not get it as a GSRect in the first place..
`let kbRect = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect,`

markj commented on PanModal/Controller/PanModalPresentationController.swift:
```diff
@@ -234,12 +246,57 @@ open class PanModalPresentationController: UIPresentationController {
+        if let keyboardSize = (notification.userInfo?  [UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue {
+            
+            if let scrollView = presentable?.panScrollable, pannedToMax {
+                scrollView.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: keyboardSize.height, right: 0.0);
```
This makes assumptions that there is no content inset for the scrollview in the first place.
You should be respecting what values are there already.

markj commented:
I'm on the outside of this.. but associating this keyboard behaviour with something that MUST have a scrollview seems wrong.. what if you present a small VC that has two fields in but doesn't need a scrollview?

Wouldn't it be better to handle the keyboard position in dealing with where the child VC is presenting. Then you aren't modifying things you aren't in charge of (e.g. scrollview content insets)

markj commented on PanModal/Controller/PanModalPresentationController.swift:
```diff
@@ -234,12 +246,57 @@ open class PanModalPresentationController: UIPresentationController {
 public extension PanModalPresentationController {
+    
+    func observeKeyboard() {
+        NotificationCenter.default.addObserver(self, selector: #selector(self.keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
```
You need to unsubscribe from these.

markj replied:
I'm wrong... you don't need to anymore.. I thought that was only the token version.

1lyyfe committed:
+ added support to keep current scroll inset and add keyboard height as opposed to disregarding the current scoll inset if there was any

1lyyfe committed:
+ always need to reset keyboard height

markj commented on PanModal/Controller/PanModalPresentationController.swift:
```diff
@@ -234,12 +256,64 @@ open class PanModalPresentationController: UIPresentationController {
+            
+            if let scrollView = presentable?.panScrollable, pannedToMax {
+                currentScrollContentInset = scrollView.contentInset.bottom + keyboardHeight
+                scrollView.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: currentScrollContentInset, right: 0.0);
```
This shouldn't have a semi colon.

you should run swift lint

markj commented:
I *think* you need to be updating the scrollbar insets as well as the content insets.

1lyyfe committed:
+ added scroll indicator support
+ removed bug of continuously adding to the scroll inset on invoking the keyboard show notification
